File: c/src/stack.c
#define _GNU_SOURCE
#include "zeroipc.h"
#include "zeroipc_stack.h"
#include <sched.h>
#include <stdatomic.h>
#include <string.h>
#include <stdlib.h>
#include <stdint.h>

/*
 * Stack binary layout (matches C++/Go/Python, format v2):
 *   [top:i32][capacity:u32][elem_size:u32][reserved:u32]  (16 bytes)
 *   [data[0]][data[1]]...[data[cap-1]]
 *   [pad to 8-byte boundary]
 *   [state[0]:u32][state[1]:u32]...[state[cap-1]:u32]
 *
 * top is signed int32: -1 = empty, otherwise index of top element.
 * Per-slot 4-state CAS protocol: EMPTY(0) -> WRITING(1) -> READY(2) -> READING(3) -> EMPTY(0)
 * The 16-byte header makes the data array 8-aligned; the state array is placed
 * on the next 8-byte boundary so its atomics are always naturally aligned.
 */

/* Round n up to the next multiple of 8 (8-byte section alignment, format v2). */
#define ZIPC_ALIGN8(n) (((n) + 7u) & ~(size_t)7u)

/* Per-slot states */
#define SLOT_EMPTY   0
#define SLOT_WRITING 1
#define SLOT_READY   2
#define SLOT_READING 3

/* Bound on every slot-state spin loop. A crashed peer leaves its slot
 * permanently claimed; bailing out (with a best-effort undo of the top
 * reservation) makes push/pop/top best-effort instead of hanging. */
#define ZIPC_MAX_SPINS 10000

/* Stack header in shared memory — matches C++ Stack::Header */
typedef struct {
    _Atomic int32_t top;       /* index of top element, -1 when empty */
    uint32_t capacity;
    uint32_t elem_size;
    uint32_t reserved;         /* pads header to 16 bytes so the data array is 8-aligned */
} stack_header_t;

/* Stack structure (process-local handle) */
struct zeroipc_stack {
    zeroipc_memory_t* memory;
    stack_header_t* header;
    void* data;                  /* pointer to data array */
    _Atomic uint32_t* state;     /* pointer to per-slot state array */
    char name[32];
};

/* Create stack */
zeroipc_stack_t* zeroipc_stack_create(zeroipc_memory_t* mem, const char* name,
                                       size_t elem_size, size_t capacity) {
    if (!mem || !name || elem_size == 0 || capacity == 0) {
        return NULL;
    }

    zeroipc_stack_t* stack = calloc(1, sizeof(zeroipc_stack_t));
    if (!stack) return NULL;

    stack->memory = mem;
    strncpy(stack->name, name, sizeof(stack->name) - 1);

    /* Layout: [header(16)][data: elem_size*capacity][pad][state: uint32*capacity] */
    size_t state_array_size = sizeof(uint32_t) * capacity;
    if (capacity > (SIZE_MAX - sizeof(stack_header_t) - state_array_size) / elem_size) {
        free(stack);
        return NULL;
    }
    size_t state_off = ZIPC_ALIGN8(elem_size * capacity);
    size_t total_size = sizeof(stack_header_t) + state_off + state_array_size;

    size_t offset;
    int result = zeroipc_table_add(mem, name, total_size, &offset);
    if (result != ZEROIPC_OK) {
        free(stack);
        return NULL;
    }

    stack->header = (stack_header_t*)((char*)zeroipc_memory_base(mem) + offset);
    stack->data = (char*)stack->header + sizeof(stack_header_t);
    stack->state = (_Atomic uint32_t*)((char*)stack->data + state_off);

    /* Initialize header (top = -1 means empty) */
    atomic_store(&stack->header->top, -1);
    stack->header->capacity = capacity;
    stack->header->elem_size = elem_size;
    stack->header->reserved = 0;

    /* Initialize per-slot states to EMPTY */
    for (uint32_t i = 0; i < capacity; ++i) {
        atomic_store_explicit(&stack->state[i], SLOT_EMPTY, memory_order_relaxed);
    }

    return stack;
}

/* Open existing stack */
zeroipc_stack_t* zeroipc_stack_open(zeroipc_memory_t* mem, const char* name) {
    if (!mem || !name) return NULL;

    size_t offset, size;
    int result = zeroipc_table_find(mem, name, &offset, &size);
    if (result != ZEROIPC_OK) return NULL;

    zeroipc_stack_t* stack = calloc(1, sizeof(zeroipc_stack_t));
    if (!stack) return NULL;

    stack->memory = mem;
    strncpy(stack->name, name, sizeof(stack->name) - 1);

    stack->header = (stack_header_t*)((char*)zeroipc_memory_base(mem) + offset);
    if (stack->header->capacity == 0 || stack->header->elem_size == 0) {
        free(stack);
        return NULL;
    }

    stack->data = (char*)stack->header + sizeof(stack_header_t);
    stack->state = (_Atomic uint32_t*)((char*)stack->data + ZIPC_ALIGN8(
        (size_t)stack->header->elem_size * stack->header->capacity));

    return stack;
}

void zeroipc_stack_close(zeroipc_stack_t* stack) {
    free(stack);
}

/* Push (lock-free with per-slot 4-state CAS) */
int zeroipc_stack_push(zeroipc_stack_t* stack, const void* value) {
    if (!stack || !value) return ZEROIPC_ERROR_SIZE;

    int32_t current_top, new_top;

    /* Step 1: Reserve slot by CAS-advancing top */
    do {
        current_top = atomic_load_explicit(&stack->header->top, memory_order_relaxed);
        if (current_top >= (int32_t)(stack->header->capacity - 1)) {
            return ZEROIPC_ERROR_SIZE;  /* full */
        }
        new_top = current_top + 1;
    } while (!atomic_compare_exchange_weak_explicit(
                &stack->header->top, &current_top, new_top,
                memory_order_acq_rel, memory_order_relaxed));

    /* Step 2: CAS slot EMPTY -> WRITING. Bounded spin: a crashed peer can
     * leave the slot permanently claimed; bail out rather than hang. */
    uint32_t expected = SLOT_EMPTY;
    int claimed = 0;
    for (int spins = 0; spins < ZIPC_MAX_SPINS; ++spins) {
        if (atomic_compare_exchange_weak_explicit(
                &stack->state[new_top], &expected, SLOT_WRITING,
                memory_order_acq_rel, memory_order_relaxed)) {
            claimed = 1;
            break;
        }
        expected = SLOT_EMPTY;
        sched_yield();
    }
    if (!claimed) {
        /* Best-effort undo of the top reservation; if another push already
         * built on top of ours the CAS fails and the slot stays orphaned. */
        int32_t reserved = new_top;
        atomic_compare_exchange_strong_explicit(
            &stack->header->top, &reserved, current_top,
            memory_order_acq_rel, memory_order_relaxed);
        return ZEROIPC_ERROR_TIMEOUT;
    }

    /* Step 3: Write data */
    void* slot = (char*)stack->data + new_top * stack->header->elem_size;
    memcpy(slot, value, stack->header->elem_size);

    /* Step 4: WRITING -> READY */
    atomic_store_explicit(&stack->state[new_top], SLOT_READY, memory_order_release);

    return ZEROIPC_OK;
}

/* Pop (lock-free with per-slot 4-state CAS) */
int zeroipc_stack_pop(zeroipc_stack_t* stack, void* value) {
    if (!stack || !value) return ZEROIPC_ERROR_SIZE;

    int32_t current_top, new_top;

    /* Step 1: Reserve slot by CAS-decrementing top */
    do {
        current_top = atomic_load_explicit(&stack->header->top, memory_order_relaxed);
        if (current_top < 0) {
            return ZEROIPC_ERROR_NOT_FOUND;  /* empty */
        }
        new_top = current_top - 1;
    } while (!atomic_compare_exchange_weak_explicit(
                &stack->header->top, &current_top, new_top,
                memory_order_acq_rel, memory_order_relaxed));

    /* Step 2: CAS slot READY -> READING. Bounded spin: normally this only
     * waits for the owning push to finish writing, but a pusher that crashed
     * mid-write leaves the slot stuck in WRITING forever. */
    uint32_t expected = SLOT_READY;
    int claimed = 0;
    for (int spins = 0; spins < ZIPC_MAX_SPINS; ++spins) {
        if (atomic_compare_exchange_weak_explicit(
                &stack->state[current_top], &expected, SLOT_READING,
                memory_order_acq_rel, memory_order_relaxed)) {
            claimed = 1;
            break;
        }
        expected = SLOT_READY;
        sched_yield();
    }
    if (!claimed) {
        /* Best-effort undo: put the item back under top so it is not
         * silently dropped. */
        int32_t reserved = new_top;
        atomic_compare_exchange_strong_explicit(
            &stack->header->top, &reserved, current_top,
            memory_order_acq_rel, memory_order_relaxed);
        return ZEROIPC_ERROR_TIMEOUT;
    }

    /* Step 3: Read data */
    void* slot = (char*)stack->data + current_top * stack->header->elem_size;
    memcpy(value, slot, stack->header->elem_size);

    /* Step 4: READING -> EMPTY */
    atomic_store_explicit(&stack->state[current_top], SLOT_EMPTY, memory_order_release);

    return ZEROIPC_OK;
}

/* Peek at top without removing.
 *
 * Best-effort: the peek must win a CAS on the slot state to read safely, so
 * under heavy contention (or a crashed peer holding the slot) it can return
 * ZEROIPC_ERROR_NOT_FOUND even though the stack is non-empty. Do not treat
 * that as an authoritative emptiness check; use zeroipc_stack_size for that. */
int zeroipc_stack_top(zeroipc_stack_t* stack, void* value) {
    if (!stack || !value) return ZEROIPC_ERROR_SIZE;

    /* A peek cannot passively read the slot: between observing READY and the
     * memcpy, a concurrent pop could recycle the slot and a new push begin
     * overwriting it, racing the read (TOCTOU). Claim the slot exclusively via
     * the same state machine pop uses (READY -> READING), copy, then restore
     * it to READY. The bounded spin preserves crash-safety. */
    for (int spins = 0; spins < ZIPC_MAX_SPINS; ++spins) {
        int32_t top = atomic_load_explicit(&stack->header->top, memory_order_acquire);
        if (top < 0) return ZEROIPC_ERROR_NOT_FOUND;

        uint32_t expected = SLOT_READY;
        if (atomic_compare_exchange_strong_explicit(
                &stack->state[top], &expected, SLOT_READING,
                memory_order_acq_rel, memory_order_relaxed)) {
            void* slot = (char*)stack->data + top * stack->header->elem_size;
            memcpy(value, slot, stack->header->elem_size);
            atomic_store_explicit(&stack->state[top], SLOT_READY, memory_order_release);
            return ZEROIPC_OK;
        }
        sched_yield();
    }
    return ZEROIPC_ERROR_NOT_FOUND;  /* Timed out */
}

int zeroipc_stack_empty(zeroipc_stack_t* stack) {
    if (!stack) return 1;
    return atomic_load_explicit(&stack->header->top, memory_order_acquire) < 0;
}

int zeroipc_stack_full(zeroipc_stack_t* stack) {
    if (!stack) return 1;
    return atomic_load_explicit(&stack->header->top, memory_order_acquire)
           >= (int32_t)(stack->header->capacity - 1);
}

size_t zeroipc_stack_size(zeroipc_stack_t* stack) {
    if (!stack) return 0;
    int32_t top = atomic_load_explicit(&stack->header->top, memory_order_acquire);
    return top < 0 ? 0 : (size_t)(top + 1);
}

size_t zeroipc_stack_capacity(zeroipc_stack_t* stack) {
    return stack ? stack->header->capacity : 0;
}

File: go/zeroipc/stack.go
package zeroipc

import (
	"encoding/binary"
	"errors"
	"fmt"
	"runtime"
	"sync/atomic"
	"unsafe"
)

// StackHeaderSize is the size of the stack header in bytes.
// Layout: top(4) + capacity(4) + elem_size(4) + reserved(4) = 16 bytes
const StackHeaderSize = 16

// align8 rounds n up to the next multiple of 8 (8-byte section alignment,
// format v2). Atomic side-arrays start on an 8-byte boundary so their atomics
// are always naturally aligned regardless of element size.
func align8(n int) int { return (n + 7) &^ 7 }

// Per-slot states for the 4-state CAS protocol (matching C++):
//
//	EMPTY(0) -> WRITING(1) -> READY(2) -> READING(3) -> EMPTY(0)
//
// Push: CAS(top), CAS(EMPTY -> WRITING), write data, store(READY)
// Pop:  CAS(top), CAS(READY -> READING), read data, store(EMPTY)
const (
	slotEmpty   uint32 = 0
	slotWriting uint32 = 1
	slotReady   uint32 = 2
	slotReading uint32 = 3
)

// Stack is a lock-free LIFO stack in shared memory.
//
// Binary layout (matching C++):
//   - top: int32 (atomic, offset 0) - index of top element, -1 when empty
//   - capacity: uint32 (offset 4)
//   - elem_size: uint32 (offset 8)
//   - data: capacity * elem_size bytes (offset 12)
//   - state: capacity * 4 bytes (after data) - per-slot atomic state
type Stack[T Numeric] struct {
	memory   *Memory
	name     string
	offset   int
	capacity uint32
	elemSize uint32
}

// NewStack creates a new stack in shared memory.
func NewStack[T Numeric](memory *Memory, name string, capacity int) (*Stack[T], error) {
	if len(name) >= NameSize {
		return nil, errors.New("name too long (max 31 characters)")
	}

	if capacity <= 0 {
		return nil, errors.New("capacity must be positive")
	}

	var zero T
	elemSize := int(unsafe.Sizeof(zero))

	// Check if already exists
	if entry := memory.Find(name); entry != nil {
		return nil, fmt.Errorf("stack '%s' already exists", name)
	}

	// Layout: [Header(16)][data: T*capacity][pad][state: uint32*capacity]
	totalSize := StackHeaderSize + align8(capacity*elemSize) + capacity*4
	offset, err := memory.Allocate(name, totalSize)
	if err != nil {
		return nil, fmt.Errorf("failed to allocate: %w", err)
	}

	data := memory.Data()

	// Write header (top = -1 means empty)
	binary.LittleEndian.PutUint32(data[offset:], 0xFFFFFFFF)          // top = -1
	binary.LittleEndian.PutUint32(data[offset+4:], uint32(capacity))  // capacity
	binary.LittleEndian.PutUint32(data[offset+8:], uint32(elemSize))  // elem_size
	binary.LittleEndian.PutUint32(data[offset+12:], 0)               // reserved

	// Zero-initialize data, padding, and state areas (state[i] = EMPTY = 0)
	start := offset + StackHeaderSize
	end := start + align8(capacity*elemSize) + capacity*4
	for i := start; i < end; i++ {
		data[i] = 0
	}

	return &Stack[T]{
		memory:   memory,
		name:     name,
		offset:   offset,
		capacity: uint32(capacity),
		elemSize: uint32(elemSize),
	}, nil
}

// OpenStack opens an existing stack in shared memory.
func OpenStack[T Numeric](memory *Memory, name string) (*Stack[T], error) {
	entry := memory.Find(name)
	if entry == nil {
		return nil, fmt.Errorf("stack '%s' not found", name)
	}

	var zero T
	elemSize := int(unsafe.Sizeof(zero))

	offset := int(entry.Offset)
	data := memory.Data()

	// Read header
	capacity := binary.LittleEndian.Uint32(data[offset+4:])
	storedElemSize := binary.LittleEndian.Uint32(data[offset+8:])

	if int(storedElemSize) != elemSize {
		return nil, fmt.Errorf("element size mismatch: stored %d, expected %d", storedElemSize, elemSize)
	}

	return &Stack[T]{
		memory:   memory,
		name:     name,
		offset:   offset,
		capacity: capacity,
		elemSize: uint32(elemSize),
	}, nil
}

// topPtr returns a pointer to the atomic top counter.
func (s *Stack[T]) topPtr() *int32 {
	return (*int32)(unsafe.Pointer(&s.memory.Data()[s.offset]))
}

// slotStatePtr returns a pointer to the atomic state for slot i.
// State array lives after the data array.
func (s *Stack[T]) slotStatePtr(i int32) *uint32 {
	stateOffset := s.offset + StackHeaderSize + align8(int(s.capacity)*int(s.elemSize)) + int(i)*4
	return (*uint32)(unsafe.Pointer(&s.memory.Data()[stateOffset]))
}

// Push adds an element to the top of the stack (lock-free with per-slot state).
// Returns false if the stack is full.
func (s *Stack[T]) Push(value T) bool {
	topPtr := s.topPtr()

	// Step 1: Reserve a slot by CAS-advancing top
	var currentTop, newTop int32
	for {
		currentTop = atomic.LoadInt32(topPtr)
		if currentTop >= int32(s.capacity)-1 {
			return false // full
		}
		newTop = currentTop + 1
		if atomic.CompareAndSwapInt32(topPtr, currentTop, newTop) {
			break
		}
	}

	// Step 2: CAS the slot state EMPTY -> WRITING
	statePtr := s.slotStatePtr(newTop)
	for {
		expected := slotEmpty
		if atomic.CompareAndSwapUint32(statePtr, expected, slotWriting) {
			break
		}
		runtime.Gosched()
	}

	// Step 3: Write data (we have exclusive ownership)
	dataOffset := s.offset + StackHeaderSize + int(newTop)*int(s.elemSize)
	*(*T)(unsafe.Pointer(&s.memory.Data()[dataOffset])) = value

	// Step 4: Publish: WRITING -> READY
	atomic.StoreUint32(statePtr, slotReady)
	return true
}

// Pop removes and returns the top element from the stack (lock-free with per-slot state).
// Returns the value and true if successful, or zero value and false if empty.
func (s *Stack[T]) Pop() (T, bool) {
	var zero T
	topPtr := s.topPtr()

	// Step 1: Reserve a slot by CAS-decrementing top
	var currentTop, newTop int32
	for {
		currentTop = atomic.LoadInt32(topPtr)
		if currentTop < 0 {
			return zero, false // empty
		}
		newTop = currentTop - 1
		if atomic.CompareAndSwapInt32(topPtr, currentTop, newTop) {
			break
		}
	}

	// Step 2: CAS the slot state READY -> READING
	statePtr := s.slotStatePtr(currentTop)
	for {
		expected := slotReady
		if atomic.CompareAndSwapUint32(statePtr, expected, slotReading) {
			break
		}
		runtime.Gosched()
	}

	// Step 3: Read data (we have exclusive ownership)
	dataOffset := s.offset + StackHeaderSize + int(currentTop)*int(s.elemSize)
	value := *(*T)(unsafe.Pointer(&s.memory.Data()[dataOffset]))

	// Step 4: Release: READING -> EMPTY
	atomic.StoreUint32(statePtr, slotEmpty)
	return value, true
}

// Top returns the top element without removing it.
// Returns the value and true if successful, or zero value and false if empty.
func (s *Stack[T]) Top() (T, bool) {
	var zero T

	// A peek cannot passively read the slot: between observing READY and the
	// read, a concurrent pop could recycle the slot and a new push begin
	// overwriting it, racing the read (TOCTOU). Claim the slot exclusively via
	// the same state machine pop uses (READY -> READING), copy, then restore it
	// to READY. The bounded spin preserves crash-safety.
	for spins := 0; spins < 10000; spins++ {
		currentTop := atomic.LoadInt32(s.topPtr())
		if currentTop < 0 {
			return zero, false
		}

		statePtr := s.slotStatePtr(currentTop)
		if atomic.CompareAndSwapUint32(statePtr, slotReady, slotReading) {
			dataOffset := s.offset + StackHeaderSize + int(currentTop)*int(s.elemSize)
			value := *(*T)(unsafe.Pointer(&s.memory.Data()[dataOffset]))
			atomic.StoreUint32(statePtr, slotReady)
			return value, true
		}
		runtime.Gosched()
	}
	return zero, false
}

// Empty returns true if the stack appears empty.
func (s *Stack[T]) Empty() bool {
	return atomic.LoadInt32(s.topPtr()) < 0
}

// Full returns true if the stack appears full.
func (s *Stack[T]) Full() bool {
	return atomic.LoadInt32(s.topPtr()) >= int32(s.capacity)-1
}

// Size returns the approximate number of elements in the stack.
func (s *Stack[T]) Size() int {
	top := atomic.LoadInt32(s.topPtr())
	if top < 0 {
		return 0
	}
	return int(top) + 1
}

// Capacity returns the maximum number of elements the stack can hold.
func (s *Stack[T]) Capacity() int {
	return int(s.capacity)
}

// Name returns the stack name.
func (s *Stack[T]) Name() string {
	return s.name
}
